File: src/lib/commands/git.ts
import type {
  BranchInfo,
  FileStatus,
  RepoInfo,
  TrackingInfo,
  WorktreeInfo,
  WorktreePreview,
} from '../../types/git';
import { apiDelete, apiGet, apiPost } from '../api/client';

export async function addRepository(): Promise<RepoInfo | null> {
  try {
    return await apiPost<RepoInfo | null>('/repos');
  } catch (e) {
    console.error(`addRepository failed: ${e}`);
    throw e;
  }
}

export async function listRepositories(): Promise<RepoInfo[]> {
  try {
    const detail = await apiGet<{ repo: RepoInfo; worktrees: WorktreeInfo[] }>('/repos');
    return [detail.repo];
  } catch (e) {
    console.error(`listRepositories failed: ${e}`);
    throw e;
  }
}

export async function removeRepository(repoPath: string): Promise<void> {
  try {
    await apiDelete(`/repos?repoPath=${encodeURIComponent(repoPath)}`);
  } catch (e) {
    console.error(`removeRepository failed: ${e}`);
    throw e;
  }
}

export async function listWorktrees(_repoPath: string): Promise<WorktreeInfo[]> {
  try {
    const detail = await apiGet<{ repo: RepoInfo; worktrees: WorktreeInfo[] }>('/repos');
    return detail.worktrees;
  } catch (e) {
    console.error(`listWorktrees failed: ${e}`);
    throw e;
  }
}

export async function createWorktree(
  repoPath: string,
  name: string,
  branch?: string,
  baseBranch?: string,
): Promise<WorktreeInfo> {
  try {
    return await apiPost<WorktreeInfo>('/repos/worktrees', {
      repoPath,
      name,
      branch,
      baseBranch,
    });
  } catch (e) {
    console.error(`createWorktree failed: ${e}`);
    throw e;
  }
}

export async function removeWorktree(
  repoPath: string,
  worktreeName: string,
  deleteBranch: boolean,
): Promise<void> {
  try {
    await apiDelete(
      `/repos/worktrees/${encodeURIComponent(worktreeName)}?repoPath=${encodeURIComponent(repoPath)}&deleteBranch=${deleteBranch}`,
    );
  } catch (e) {
    console.error(`removeWorktree failed: ${e}`);
    throw e;
  }
}

export async function previewWorktree(repoPath: string, name: string): Promise<WorktreePreview> {
  try {
    return await apiGet<WorktreePreview>(
      `/repos/worktrees/${encodeURIComponent(name)}/preview?repoPath=${encodeURIComponent(repoPath)}`,
    );
  } catch (e) {
    console.error(`previewWorktree failed: ${e}`);
    throw e;
  }
}

export async function listBranches(repoPath: string): Promise<BranchInfo[]> {
  try {
    return await apiGet<BranchInfo[]>(`/repos/branches?repoPath=${encodeURIComponent(repoPath)}`);
  } catch (e) {
    console.error(`listBranches failed: ${e}`);
    throw e;
  }
}

export async function getBranchTracking(
  repoPath: string,
  branchName: string,
): Promise<TrackingInfo | null> {
  try {
    return await apiGet<TrackingInfo | null>(
      `/repos/branches/${encodeURIComponent(branchName)}/tracking?repoPath=${encodeURIComponent(repoPath)}`,
    );
  } catch (e) {
    console.error(`getBranchTracking failed: ${e}`);
    throw e;
  }
}

export async function getRepoStatus(worktreePath: string): Promise<FileStatus[]> {
  try {
    return await apiGet<FileStatus[]>(
      `/repos/status?worktreePath=${encodeURIComponent(worktreePath)}`,
    );
  } catch (e) {
    console.error(`getRepoStatus failed: ${e}`);
    throw e;
  }
}

File: src-tauri/src/commands/git.rs
use crate::error::AppError;
use crate::models::{BranchInfo, FileStatus, RepoInfo, WorktreeInfo, WorktreePreview};
use crate::services::{config, git};
use std::path::PathBuf;
use tauri_plugin_dialog::DialogExt;

#[tauri::command]
pub async fn add_repository(app: tauri::AppHandle) -> Result<Option<RepoInfo>, AppError> {
    let folder =
        tauri::async_runtime::spawn_blocking(move || app.dialog().file().blocking_pick_folder())
            .await
            .map_err(|e| AppError::Config(e.to_string()))?;

    let Some(path) = folder else {
        return Ok(None);
    };

    let path_buf = path.as_path().ok_or_else(|| {
        AppError::Config("Selected path is not a valid filesystem path".to_string())
    })?;

    let repo_info = git::validate_repo(path_buf)?;

    let mut global_config = config::load_global_config()?;
    let path_str = repo_info.path.to_string_lossy().to_string();
    if !global_config.repos.contains(&path_str) {
        global_config.repos.push(path_str);
        config::save_global_config(&global_config)?;
    }

    Ok(Some(repo_info))
}

#[tauri::command]
pub fn list_repositories() -> Result<Vec<RepoInfo>, AppError> {
    let global_config = config::load_global_config()?;
    let mut repos = Vec::new();

    for repo_path in &global_config.repos {
        if let Ok(info) = git::validate_repo(&PathBuf::from(repo_path)) {
            repos.push(info);
        }
    }

    Ok(repos)
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn remove_repository(repo_path: String) -> Result<(), AppError> {
    let mut global_config = config::load_global_config()?;
    global_config.repos.retain(|p| p != &repo_path);
    config::save_global_config(&global_config)?;
    Ok(())
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn list_worktrees_cmd(repo_path: String) -> Result<Vec<WorktreeInfo>, AppError> {
    git::list_worktrees(&PathBuf::from(&repo_path))
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn create_worktree_cmd(
    repo_path: String,
    name: String,
    branch: Option<String>,
    base_branch: Option<String>,
) -> Result<WorktreeInfo, AppError> {
    git::create_worktree(
        &PathBuf::from(&repo_path),
        &name,
        branch.as_deref(),
        base_branch.as_deref(),
    )
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn remove_worktree_cmd(
    repo_path: String,
    worktree_name: String,
    delete_branch: bool,
) -> Result<(), AppError> {
    git::remove_worktree(&PathBuf::from(&repo_path), &worktree_name, delete_branch)
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn preview_worktree_cmd(repo_path: String, name: String) -> Result<WorktreePreview, AppError> {
    git::preview_worktree(&PathBuf::from(&repo_path), &name)
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn get_repo_status(worktree_path: String) -> Result<Vec<FileStatus>, AppError> {
    git::get_status(&PathBuf::from(&worktree_path))
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn list_branches_cmd(repo_path: String) -> Result<Vec<BranchInfo>, AppError> {
    git::list_branches(&PathBuf::from(&repo_path))
}
